Settings map

whmii opened the pull request:
Addresses https://github.com/whmii/cask/issues/7

All grid-settings have been moved to  the `$grid-settings` map. For convenience/brevity, there is also a `grid-settings()` function that can read and write values in this map.

Also, the file import order has been updated so that functions are now imported *before* variables. This has been done to allow the use of the `grid-settings()` function inside the variable file.

![screencapture-localhost-8000-1454449902405](https://cloud.githubusercontent.com/assets/2489247/12765588/53f4f2f0-c9cd-11e5-85d2-de134bf55caa.png)

whmii commented:
Attached a screenshot for the "yes, it works"

whmii commented on app/assets/stylesheets/_variables.scss:
```diff
@@ -1,22 +1,21 @@
 
-// ========= Grid Variables =========
+$grid-settings: (
+  text-direction: ltr,
```
:abc: ?… i guess?

tysongach replied:
:+1:  to :abc: the :map:

kaishin commented:
This looks good. One thing to keep in mind is that doing map operations is slower. Should be fine but could get noticeably slower if used without restraints.

tysongach commented on app/assets/stylesheets/functions/_grid-settings.scss:
```diff
@@ -0,0 +1,12 @@
+//
+// `$grid-settings: grid-settings(key, value);`
+
+@function grid-settings($key, $value: false) {
```
I think this should be broken into two functions: one to retrieve and one to update.

Right now, throughout the codebase, you see `grid-settings()` and it’s not super clear what it is doing, which can be two vastly different things.

It also eliminates the need for the `@if` statement here. #smallerdryerfunctions :smile:

tysongach replied:
Helpful reference: http://www.sitepoint.com/functional-wrappers-sass/

tysongach commented:
Love maps and this is a big improvement, but would love to see more, smaller functions rather than large ones with many arguments/options.

tysongach commented:
@whmii Can you breakdown the intended use of this, from the perspective of the user with Cask in their project?

For Bourbon, I was aiming for this:

``` scss
// User's main Sass manifest
@import "bourbon";
@import "variables"; // this contains project settings
@import "components";
```

``` scss
// _variables.scss
// Below, the user is overriding the default modular scale base
$bourbon: (
  "modular-scale-base": 2em,
);
```

``` scss
// _components.scss
// The output from the rule below should be based off of a
// modular scale base of 2em, since the user changed it to that
.element {
  font-size: modular-scale(1);
}
```

The above mimics exactly how Bourbon settings work today, with the exception that each setting variable is now a key/value pair in the `$bourbon` map. But placement within the project should be the same.

Is this aligned or the same with your expectations? Or totally different?

whmii commented:
Thanks for the feedback @tysongach @kaishin. I updated the PR with a new basic map function. `grid-settings()` is now read only and retrieves the value from a dynamic map that is merged, when the function is called, that returns `map-merge($grid-defaults,  $grid-settings)` :clap: :boom: 

Also added a separate `grid-settings-assign()` function that runs a map merge on grid-settings so you can update values like:

``` scss
$grid-settings: (blah: blah);
$grid-settings: grid-settings((foo: bar));

@return $grid-settings;
  (blah: blah, foo: bar)
```

whmii commented:
Works like a charm btw
